library/psa_crypto_core.h

```diff
@@ -607,4 +607,275 @@ psa_status_t psa_key_agreement_raw_builtin(
     size_t shared_secret_size,
     size_t *shared_secret_length );
 
+/**
+ * \brief Set the maximum number of ops allowed to be executed by an
+ *        interruptible function in a single call.
+ *
+ * \note The signature of this function is that of a PSA driver
+ *       interruptible_set_max_ops entry point. This function behaves as an
+ *       interruptible_set_max_ops entry point as defined in the PSA driver
+ *       interface specification for transparent drivers.
+ *
+ * \param[in]  max_ops          The maximum number of ops to be executed in a
+ *                              single call, this can be a number from 0 to
+ *                              #PSA_INTERRUPTIBLE_MAX_OPS_UNLIMITED, where 0
+ *                              is obviously the least amount of work done per
+ *                              call.
+ */
+void mbedtls_psa_interruptible_set_max_ops( uint32_t max_ops );
+
+/**
+ * \brief Get the maximum number of ops allowed to be executed by an
+ *        interruptible function in a single call.
+ *
+ * \note The signature of this function is that of a PSA driver
+ *       interruptible_get_max_ops entry point. This function behaves as an
+ *       interruptible_get_max_ops entry point as defined in the PSA driver
+ *       interface specification for transparent drivers.
+ *
+ * \return                      Maximum number of ops allowed to be executed
+ *                              by an interruptible function in a single call.
+ */
+uint32_t mbedtls_psa_interruptible_get_max_ops( void );
+
+/**
+ * \brief Get the number of ops that a hash signing operation has taken for the
+ *        previous call. If no call or work has taken place, this will return
+ *        zero.
+ *
+ * \note The signature of this function is that of a PSA driver
+ *       sign_get_num_ops entry point, however it differs in behaviour from the
+ *       driver function in that this function returns a delta of work done in
+ *       the last call rather than all of the ops done ever by the whole
+ *       operation, due to internal implementation differences.
```

Comment:
In the previous implementation I had implemented a += inside this function to hide the fact that the internal implementation returns a delta, not a an absolute number of work units, and so that drivers etc would have a 'normal' interface. Both you and Manuel agreed that this was unacceptable, so I fixed it the way you specified in the previous comment thread, where I also pointed out that this would cause oddities in the implementation, such as this.  The only option I have left is to remove num_ops from the PSA level and move it down to the mbedtls_psa_level (and presumably any driver) - that way everything returns an absolute value, however we will need another driver call for the 'full' abort, which actually clears the number of ops when the end user calls abort. Is this acceptable?

Reply:
> remove num_ops from the PSA level and move it down to the mbedtls_psa_level (and presumably any driver)

Makes sense to me.

An alternative would be to remove the `get_ops` interface, and instead give each driver operation an output parameter to be filled with the number of ops performed. But I'm not convinced that's a better interface design, and I don't want to do a significant interface change at this point.

> we will need another driver call for the 'full' abort, which actually clears the number of ops when the end user calls abort

I don't see why. When the operation is aborted, we just call the driver's abort method. The op count in the core structure doesn't involve the driver at all.

Reply:
Ah, you mean have a num_ops both at the PSA operation level and the mbedtls_psa / driver level. Ok, that is admittedly simpler to implement.

Reply:
Even if the driver keeps track of the op count, it can zero that out in its abort function, or it can leave it unzeroed and zero it in its start function. Keeping the op count available for the application is purely a core concern.